File: egarch-volatility-modeling/egarch.py
# ==============================================================================
# egarch.py  –  EGARCH(1,1) via arch
# Data      :  returns.csv  –  column "Returns"
# ==============================================================================
import pandas as pd
import numpy as np
from arch import arch_model

# ------------------------------------------------------------------------------
# 1. Load data
# ------------------------------------------------------------------------------
data    = pd.read_csv('returns.csv')
returns = data['Returns'].dropna().astype(float)

# ------------------------------------------------------------------------------
# 2. Artificial daily date index
# ------------------------------------------------------------------------------
start_date    = pd.to_datetime('2015-01-01')
returns.index = pd.date_range(start=start_date, periods=len(returns), freq='D')

# ------------------------------------------------------------------------------
# 3. Specify EGARCH(1,1) with Student-t innovations
# ------------------------------------------------------------------------------
model = arch_model(
    returns,
    vol  = 'EGARCH',
    p    = 1,
    o    = 1,   # asymmetry / leverage
    q    = 1,
    mean = 'Constant',
    dist = 't'  # Student-t for heavy tails
)

# ------------------------------------------------------------------------------
# 4. Fit
# ------------------------------------------------------------------------------
fit_py = model.fit(disp='off')
print(fit_py.summary())

# ------------------------------------------------------------------------------
# 5. One-step-ahead forecast
# ------------------------------------------------------------------------------
fc_py          = fit_py.forecast(horizon=1)
forecast_mean  = fc_py.mean.iloc[-1].values
forecast_sigma = np.sqrt(fc_py.variance.iloc[-1].values)

print()
print('[Python] Forecasted mean return (next 1 day):')
print(forecast_mean)
print('[Python] Forecasted sigma (next 1 day):')
print(forecast_sigma)

File: egarch-volatility-modeling/egarch.R
# ==============================================================================
# egarch.R  –  EGARCH(1,1) via rugarch
# Data     :  returns.csv  –  column "Returns"
# ==============================================================================
library(rugarch)
library(xts)

# ------------------------------------------------------------------------------
# 1. Load data
# ------------------------------------------------------------------------------
data    <- read.csv("returns.csv")
returns <- as.numeric(data$Returns)

# ------------------------------------------------------------------------------
# 2. Artificial daily date index
# ------------------------------------------------------------------------------
time_index  <- seq.Date(from = as.Date("2015-01-01"),
                        by   = "days",
                        length.out = length(returns))
returns_xts <- xts(returns, order.by = time_index)

# ------------------------------------------------------------------------------
# 3. Specify EGARCH(1,1) with Student-t innovations
# ------------------------------------------------------------------------------
spec <- ugarchspec(
  variance.model = list(model      = "eGARCH",
                        garchOrder = c(1, 1)),
  mean.model     = list(armaOrder  = c(0, 0),
                        include.mean = TRUE),
  distribution.model = "std"          # Student-t for heavy tails
)

# ------------------------------------------------------------------------------
# 4. Fit
# ------------------------------------------------------------------------------
fit_r <- ugarchfit(spec = spec, data = returns_xts)
show(fit_r)

# ------------------------------------------------------------------------------
# 5. One-step-ahead forecast
# ------------------------------------------------------------------------------
fc_r <- ugarchforecast(fit_r, n.ahead = 1)
cat("\n[R] Forecasted mean return (next 1 day):\n")
print(fc_r@forecast$seriesFor)
cat("\n[R] Forecasted sigma (next 1 day):\n")
print(fc_r@forecast$sigmaFor)
